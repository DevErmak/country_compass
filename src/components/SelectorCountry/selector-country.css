.label-search {
  font-size: 20px;
  font-weight: 400;
  margin-top: 48px;
  margin-left: 113px;
  margin-right: 714px;
}

.custom-select__control {
  position: relative !important;
  -webkit-transition: all 100ms !important;
  transition: all 100ms !important;
  background-color: #fff !important;
  border: none !important;
  box-shadow: none !important;
  border-radius: 10px !important;
  width: 798px !important;
  height: 46px !important;
  margin-left: 113px !important;
}
.custom-select__menu {
  position: relative !important;
  -webkit-transition: all 100ms !important;
  transition: all 100ms !important;
  background-color: #fff !important;
  border: none !important;
  box-shadow: none !important;
  color: #000000 !important;
  width: 798px !important;
  height: 46px !important;
  margin-left: 113px !important;
}

.custom-select__input {
  color: #000000 !important;
}

.custom-select__option {
  cursor: pointer !important;
  background-color: #fff !important;

  /* &--is-focused {
    background-color: #83c6fa !important;
    color: #ff0000 !important;
  } */
}
.custom-select__option:hover {
  background-color: #83c6fa !important;
  color: #293251 !important;
}
