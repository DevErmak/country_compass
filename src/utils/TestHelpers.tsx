/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable import/no-extraneous-dependencies */
import React from 'react';
import ReactDOM from 'react-dom';
import { configureStore, combineReducers } from '@reduxjs/toolkit';
import type { Store } from '@reduxjs/toolkit';
import { render } from '@testing-library/react';
import { FormProvider, useForm } from 'react-hook-form';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import infoCountriesReducer from '../store/country/infoCountrySlice';
import infoUserReducer from '../store/user/infoUserSlice';
import createSagaMiddleware from 'redux-saga';
export const saga = createSagaMiddleware();

const rootReducer = combineReducers({
  infoCountries: infoCountriesReducer,
  infoUser: infoUserReducer,
});

interface IFormWrapperProps {
  children: React.ReactNode;
  defaultValues: Record<string, any>;
}

interface IInitialEntry {
  pathname: string;
  hash?: string;
  key?: string;
  search?: string;
  state?: any;
}

const FormWrapper: React.FC<IFormWrapperProps> = ({ defaultValues, children }) => {
  const formMethods = useForm<typeof defaultValues>({
    defaultValues,
  });

  return <FormProvider<typeof defaultValues> {...formMethods}>{children}</FormProvider>;
};

type RenderComponentWithStoreArgs = {
  Component: React.FC<any>;
  children?: React.ReactNode;
  defaultFormValues?: Record<string, any>;
  memoryRouterInitialEntries?: IInitialEntry[];
  preloadedState?: Record<string, any>;
  props?: Record<string, any>;
};

export const renderComponentWithStore = ({
  Component,
  props,
  preloadedState,
  defaultFormValues,
  memoryRouterInitialEntries = [{ pathname: '/' }],
  children,
}: RenderComponentWithStoreArgs): {
  container: HTMLElement;
  mockStore: Store;
  rerender: (rerenderProps: Record<string, any>) => void;
} => {
  const mockStore = configureStore({
    reducer: rootReducer,
    middleware: [saga],
    preloadedState,
  });

  const renderComponent = (renderProps?: Record<string, any>) => (
    <Provider store={mockStore}>
      <MemoryRouter initialEntries={memoryRouterInitialEntries}>
        {defaultFormValues ? (
          <FormWrapper defaultValues={defaultFormValues}>
            <Component {...renderProps}>{children}</Component>
          </FormWrapper>
        ) : (
          <Component {...renderProps}>{children}</Component>
        )}
        <div id="portal"></div>
      </MemoryRouter>
    </Provider>
  );

  const { container, rerender: reactRerender } = render(renderComponent(props));

  const rerender = (rerenderProps: Record<string, any>) =>
    reactRerender(renderComponent(rerenderProps));

  return { container, mockStore, rerender };
};

type RenderComponentArgs = {
  Component: React.FC<any>;
  children?: React.ReactNode;
  defaultFormValues?: Record<string, any>;
  memoryRouterInitialEntries?: IInitialEntry[];
  props?: Record<string, any>;
};

export const renderComponent = ({
  Component,
  props,
  defaultFormValues,
  memoryRouterInitialEntries = [{ pathname: '/' }],
  children,
}: RenderComponentArgs): { container: HTMLElement } => {
  const { container } = render(
    <MemoryRouter initialEntries={memoryRouterInitialEntries}>
      {defaultFormValues ? (
        <FormWrapper defaultValues={defaultFormValues}>
          <Component {...props}>{children}</Component>
        </FormWrapper>
      ) : (
        <Component {...props}>{children}</Component>
      )}
      <div id="portal"></div>
    </MemoryRouter>,
  );

  return { container };
};
