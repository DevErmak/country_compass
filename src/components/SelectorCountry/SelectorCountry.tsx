import { useSelector } from 'react-redux';
import AsyncSelect from 'react-select/async';
import { getOptionsCountry } from '../../store/country/countriesSelectors';
import { IOptionCountry } from '../../store/country/types';
import { useDispatch } from 'react-redux';
import {
  getFullInfoCountryFetch,
  getListCountriesFetch,
} from '../../store/country/infoCountrySlice';
import { useEffect } from 'react';
import './selector-country.css';

type Props = {};

export default function Header({}: Props) {
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(getListCountriesFetch());
  }, []);

  const optionsCountry: IOptionCountry[] = useSelector(getOptionsCountry);

  const findCounty = (inputValue: string) => {
    return optionsCountry.filter((item) =>
      item.label.toLowerCase().includes(inputValue.toLowerCase()),
    );
  };

  let timerId: ReturnType<typeof setTimeout>;

  const promiseOptions = (inputValue: string) => {
    return new Promise<IOptionCountry[]>((resolve, reject) => {
      clearTimeout(timerId);
      timerId = setTimeout(async () => {
        try {
          const results = await findCounty(inputValue);
          resolve(results);
        } catch (error) {
          reject(error);
        }
      }, 1000);
    });
  };
  const getFullInfoCountry = (value: IOptionCountry | null) => {
    dispatch(getFullInfoCountryFetch(value?.value));
  };

  return (
    <div className="header-container">
      <div className="label-search">Search your country:</div>
      <div className="find-country">
        <AsyncSelect
          placeholder={null}
          classNamePrefix="custom-select"
          defaultOptions={optionsCountry}
          loadOptions={promiseOptions}
          components={{
            IndicatorSeparator: () => null,
            DropdownIndicator: () => null,
          }}
          onChange={getFullInfoCountry}
        />
      </div>
    </div>
  );
}
