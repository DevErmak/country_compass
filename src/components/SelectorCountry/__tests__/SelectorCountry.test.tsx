import { renderComponentWithStore } from '../../../utils/TestHelpers';
import SelectorCountry from '../SelectorCountry';

it('renders correctly selector-country', () => {
  const { container } = renderComponentWithStore({
    Component: SelectorCountry,
    preloadedState: {
      infoCountries: {
        listCountries: [],
        fullInfoCountry: [],
        isFullInfoCountry: false,
        isLoading: false,
        infoErrorResponse: '',
      },
      infoUser: {
        userName: '',
        listFavoriteCountries: [],
        isAuthentication: false,
        infoModal: { isActiveModal: false, formModal: 'login' },
      },
    },
  });
  expect(container).toMatchSnapshot();
});
