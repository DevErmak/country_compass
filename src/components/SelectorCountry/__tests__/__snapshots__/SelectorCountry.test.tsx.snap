// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly selector-country 1`] = `
<div>
  <div
    class="header-container"
  >
    <div
      class="label-search"
    >
      Search your country:
    </div>
    <div
      class="find-country"
    >
      <div
        class=" css-b62m3t-container"
      >
        <span
          class="css-1f43avz-a11yText-A11yText"
          id="react-select-2-live-region"
        />
        <span
          aria-atomic="false"
          aria-live="polite"
          aria-relevant="additions text"
          class="css-1f43avz-a11yText-A11yText"
        />
        <div
          class="custom-select__control css-13cymwt-control"
        >
          <div
            class="custom-select__value-container css-1fdsijx-ValueContainer"
          >
            <div
              class="custom-select__placeholder css-1jqq78o-placeholder"
              id="react-select-2-placeholder"
            />
            <div
              class="custom-select__input-container css-qbdosj-Input"
              data-value=""
            >
              <input
                aria-autocomplete="list"
                aria-describedby="react-select-2-placeholder"
                aria-expanded="false"
                aria-haspopup="true"
                autocapitalize="none"
                autocomplete="off"
                autocorrect="off"
                class="custom-select__input"
                id="react-select-2-input"
                role="combobox"
                spellcheck="false"
                style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                tabindex="0"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class="custom-select__indicators css-1hb7zxy-IndicatorsContainer"
          />
        </div>
      </div>
    </div>
  </div>
  <div
    id="portal"
  />
</div>
`;
